### ISSUE TITLE
`torch.nn.functional.kl_div` silently ignores wrong inputs

### ISSUE THREAD

**Issue by pmeier** [pytorch/pytorch#69230]
## 🐛 Bug

With `log_target=False` (default), `target` should be non-negative. The current implementation simply substitutes `0` for the loss in case it encounters a negative `target`:

https://github.com/pytorch/pytorch/blob/fbaa19a6fa6c7d07ecd74e5ebf3c5d0fc77ecc51/aten/src/ATen/native/Loss.cpp#L117

## To Reproduce

```python
import torch
from torch.nn.functional import kl_div

input = torch.rand(()).log()
target = torch.tensor(-1, dtype=input.dtype)

print(kl_div(input, target))
```
```
tensor(0.)
```

## Expected behavior

This is problematic for two reasons:

  1. It is mathematically wrong, since the equation (in particular `target * log(target)`) is undefined for negative targets.
  2. It most likely hides a bug in the user code.

Returning `tensor(nan)` for negative targets would resolve both issues.

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345
